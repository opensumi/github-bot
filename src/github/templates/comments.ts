import {
  renderPrOrIssueLink,
  renderPrOrIssueText,
  renderRepoLink,
  renderUserLink,
} from '.';
import { ExtractPayload } from '../template';

export function handleIssueComment(payload: ExtractPayload<'issue_comment'>) {
  const issue = payload.issue;
  const action = payload.action;
  const comment = payload.comment;
  const isUnderPullRequest = Boolean(issue.pull_request);
  let location = 'issue';
  if (isUnderPullRequest) {
    location = 'pull request';
  }
  const title = `Comment ${action} on ${location} ${renderPrOrIssueText(
    payload.issue,
  )}`;
  const text = `${renderRepoLink(payload.repository)} [Comment](${
    comment.html_url
  }) ${action} by ${renderUserLink(
    payload.sender,
  )} on ${location} ${renderPrOrIssueLink(payload.issue)}
>
> ${comment.body}
`;
  return {
    title,
    text,
  };
}
