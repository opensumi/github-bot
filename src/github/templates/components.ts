import capitalize from 'lodash/capitalize';

import { StringBuilder, limitTextByPosition } from '@/utils/string-builder';

import { render } from '../renderer';

export function RepositoryLink(repository: { name: string; html_url: string }) {
  return `[[${escapeUsername(repository.name)}]](${repository.html_url})`;
}

function escapeUsername(text: string) {
  return text.replace(/\[/g, '\\[').replace(/\]/g, '\\]');
}

export function SenderLink(sender: { login: string; html_url: string }) {
  return `[${escapeUsername(sender.login)}](${sender.html_url})`;
}

export function TeamLink(team: { name: string; html_url: string }) {
  return `[${escapeUsername(team.name)}](${team.html_url})`;
}

export function AtSenderLink(sender: { login: string; html_url: string }) {
  return `[@${escapeUsername(sender.login)}](${sender.html_url})`;
}

export function ReleaseLink(release: {
  name: string;
  tag_name: string;
  html_url: string;
}) {
  return `[${escapeUsername(release.name || release.tag_name)}](${
    release.html_url
  })`;
}

export function IssuesText(
  p: {
    /**
     * The title of the pull request.
     */
    title: string;
    /**
     * Number uniquely identifying the pull request within its repository.
     */
    number?: number;
  },
  prefix?: string,
  suffix?: string,
) {
  let text = `${prefix ?? ''}`;
  if (p.number) {
    text += `#${p.number} `;
  }
  text += p.title;
  if (suffix) {
    text += suffix;
  }
  return text;
}

export function IssuesLink(
  p: {
    /**
     * The title of the pull request.
     */
    title: string;
    /**
     * Number uniquely identifying the pull request within its repository.
     */
    number?: number;
    html_url: string;
  },
  prefix?: string,
  suffix?: string,
) {
  return `[${IssuesText(p, prefix, suffix)}](${p.html_url})`;
}

export function IssuesTitleLink(p: {
  /**
   * The title of the pull request.
   */
  title: string;
  /**
   * Number uniquely identifying the pull request within its repository.
   */
  number: number;
  html_url: string;
  body?: string | null | undefined;
}) {
  return `#### ${IssuesLink(p)}`;
}

export function PullRequestRefInfo(data: {
  head: {
    label: string;
  };
  base: {
    ref: string;
    user: {
      login: string;
    };
  };
}) {
  // display PR related info, such as pr assignees, base branch, head branch, etc.
  const head = data.head;
  const base = data.base;
  const headLabel = removeOrgInfo(base.user.login, head.label);
  return `${base.ref} <- ${headLabel}  `;
}

export function AssigneesInfo(
  assignees: {
    login: string;
    html_url: string;
  }[],
) {
  const assigneeNames = assignees.map((v) => SenderLink(v)).join(', ');
  return `Assignees: ${assigneeNames}  `;
}

interface ISenderBasic {
  login: string;
  html_url: string;
}

interface ITeamBasic {
  name: string;
  html_url: string;
}

export function RequestedReviewersInfo(
  reviewers: (ISenderBasic | ITeamBasic)[],
) {
  const reviewerNames = reviewers
    .map((v) =>
      (v as ISenderBasic).login
        ? SenderLink(v as ISenderBasic)
        : TeamLink(v as ITeamBasic),
    )
    .join(', ');
  return `Requested reviewers: ${reviewerNames}  `;
}

export function DeletedPrOrIssueTitleLink(p: {
  /**
   * The title of the pull request.
   */
  title: string;
  /**
   * Number uniquely identifying the pull request within its repository.
   */
  number: number;
  html_url: string;
  body: string | null | undefined;
}) {
  return `#### ${IssuesLink(p, '~~', '~~')}`;
}

export function Reference(text?: string | null | undefined, bodyLimit = -1) {
  if (!text) {
    return '';
  }

  if (bodyLimit && bodyLimit > 0) {
    text = limitTextByPosition(text, bodyLimit);
  }

  const arrayOfLines = text.replace(/\r\n|\n\r|\n|\r/g, '\n').split('\n');
  const newLines = [];

  for (const line of arrayOfLines) {
    if (line.trim()) {
      newLines.push(`> ${line}`);
    } else {
      newLines.push(`>`);
    }
  }

  return newLines.join('\n');
}

export class StopHandleError extends Error {
  constructor(reason: string) {
    super(reason);
  }
}

export type TextTplInput = {
  payload: any;
  title: string;
  target?: string;
  compactTitle?: string;
  body?: string | null;
  event: string;
  action?: string;
  contentLimit?: number;
  notCapitalizeTitle?: boolean;
  doNotRenderBody?: boolean;
  autoRef?: boolean;

  blockUsers?: Set<string>;
  blockActions?: Set<string>;
};

export type TemplateRenderResult = {
  title: string;
  text: string;
  compactText?: string;
};

type TextTpl = (
  data: TextTplInput,
  ctx?: {
    setting?: {
      notDisplayRepoName?: boolean;
    };
  },
) => TemplateRenderResult;

export type HandlerResult = {
  text: TextTplInput;
};

export const depManageBotToIgnore = new Set([
  'renovate[bot]',
  'dependabot[bot]',
]);

export const Template: TextTpl = (data, ctx) => {
  const {
    payload,
    title: bodyTitle,
    compactTitle,
    target,
    body,
    notCapitalizeTitle,
    autoRef,
    contentLimit = -1,
    blockActions,
    event,
    blockUsers,
  } = data;
  const action = data.action ?? payload.action;

  if (!action) {
    throw new Error('action is required for event ' + event);
  }

  if (blockActions && blockActions.has(action)) {
    throw new StopHandleError(
      `skip event ${event} action ${action} because it is in blockActions`,
    );
  }

  if (blockUsers && payload.sender && blockUsers.has(payload.sender.login)) {
    throw new StopHandleError(
      `skip event ${event} action ${action} because it is in blockUsers`,
    );
  }

  const repo = payload.repository;

  let repoInfo = RepositoryLink(repo) + ' ';
  if (ctx?.setting?.notDisplayRepoName) {
    repoInfo = '';
  }

  const textBuilder = new StringBuilder(
    `#### ${repoInfo}${bodyTitle.trim()}  `,
  );
  let compactTextBuilder: StringBuilder | undefined;
  if (compactTitle) {
    compactTextBuilder = new StringBuilder(compactTitle);
  }

  if (target) {
    textBuilder.add(target);
    // compact text do not need target
  }

  let bodyText = '';
  if (body && !data.doNotRenderBody) {
    bodyText = render(body.trim(), payload);
  }

  if (bodyText) {
    textBuilder.addDivider('', true);

    if (contentLimit && contentLimit > 0) {
      bodyText = limitTextByPosition(bodyText, contentLimit);
    }

    payload.bodyText = bodyText;

    const refText = autoRef ? '{{bodyText|ref}}' : '{{bodyText}}';
    textBuilder.add(refText);
    compactTextBuilder && compactTextBuilder.add(refText);
  }

  let titleText = `${notCapitalizeTitle ? event : capitalize(event)} ${action}`;
  if (!ctx?.setting?.notDisplayRepoName) {
    titleText = `[{{repository.name}}] ${titleText}`;
  }

  const titleBuilder = new StringBuilder(titleText);

  return {
    title: titleBuilder.render(payload),
    text: textBuilder.render(payload),
    compactText: compactTextBuilder
      ? compactTextBuilder.render(payload)
      : undefined,
  };
};

const removeOrgInfo = (orgName: string, label: string) => {
  const prefix = `${orgName}:`;
  if (label.startsWith(prefix)) {
    return label.slice(prefix.length);
  }
  return label;
};

/**
 * `not_planned` -> `not planned`
 */
export function prettyUnderlineWord(stateReason: string) {
  return stateReason.replace(/_/g, ' ');
}
