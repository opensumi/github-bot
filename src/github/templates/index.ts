export * from './prOrIssue';
export * from './comment';
export * from './release';
export * from './review';
export * from './utils';

import {
  handlePr,
  handleIssue,
  handleReview,
  handleRelease,
  handleDiscussion,
  handleIssueComment,
  handleDiscussionComment,
} from '.';

import { EmitterWebhookEventName } from '@octokit/webhooks/dist-types/types';
import { TemplateMapping } from '@/github/types';

export const templates = {
  'issues.opened': handleIssue,
  'issues.closed': handleIssue,
  'issues.reopened': handleIssue,
  'issues.edited': handleIssue,
  'pull_request.opened': handlePr,
  'pull_request.reopened': handlePr,
  'pull_request.closed': handlePr,
  'pull_request.edited': handlePr,
  'pull_request.ready_for_review': handlePr,
  'discussion.created': handleDiscussion,
  'discussion.edited': handleDiscussion,
  'discussion.deleted': handleDiscussion,
  'discussion_comment.created': handleDiscussionComment,
  'discussion_comment.deleted': handleDiscussionComment,
  'issue_comment.created': handleIssueComment,
  'issue_comment.deleted': handleIssueComment,
  'release.published': handleRelease,
  'release.released': handleRelease,
  'pull_request_review.submitted': handleReview,
  'pull_request_review.edited': handleReview,
  'pull_request_review.dismissed': handleReview,
} as TemplateMapping;

export const supportTemplates = Object.keys(
  templates,
) as EmitterWebhookEventName[];
