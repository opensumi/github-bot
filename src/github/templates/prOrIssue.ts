import { Issue, PullRequest, Discussion } from '@octokit/webhooks-types';

import { StringBuilder } from '@/utils/string-builder';

import { ExtractPayload, THasChanges, Context } from '../types';

import {
  IssuesTitleLink,
  DeletedPrOrIssueTitleLink,
  StopHandleError,
  PullRequestRefInfo,
  AssigneesInfo,
  RequestedReviewersInfo,
  Template,
  prettyUnderlineWord,
  TemplateRenderResult,
} from './components';

export type Name = 'issues' | 'pull_request' | 'discussion';
export const NameBlock = {
  issues: 'issue',
  pull_request: 'pull request',
  discussion: 'discussion',
} as {
  [key in Name]: string;
};

function render(
  name: Name,
  payload: ExtractPayload<Name>,
  data: Issue | Discussion,
  ctx: Context,
): TemplateRenderResult {
  const nameBlock = NameBlock[name];
  const action = payload.action as string;

  let shouldRenderAssigneeInfo = false;
  if (['opened', 'edited'].includes(action)) {
    shouldRenderAssigneeInfo = true;
  }

  let shouldRenderBody = true;
  if (['closed', 'edited'].includes(action)) {
    shouldRenderBody = false;
  }

  const builder = new StringBuilder();

  builder.add(IssuesTitleLink(data));

  let contentLimit = ctx.setting.contentLimit;

  if (name === 'issues') {
    // if the issue has label `monthly-report`, we will render all content
    if (
      (data as Issue).labels?.find(
        (v) => v.name && v.name.includes('monthly-report'),
      )
    ) {
      contentLimit = -1;
    }

    if (shouldRenderAssigneeInfo && (data as Issue).assignees?.length) {
      builder.add('> ' + AssigneesInfo((data as Issue).assignees));
    }
  }

  let title = `{{sender | link:sender}} ${action} [${nameBlock}](${data.html_url})`;

  if ((data as Issue).state_reason) {
    title += ` as ${prettyUnderlineWord((data as Issue).state_reason!)}`;
  }

  return Template(
    {
      payload,
      event: `${nameBlock}#${data.number}`,
      action,
      title,
      target: builder.build(),
      body: data.body,
      contentLimit,
      doNotRenderBody: !shouldRenderBody,
    },
    ctx,
  );
}

export async function handlePr(
  payload: ExtractPayload<'pull_request'>,
  ctx: Context,
) {
  const nameBlock = 'pull request';
  const data = payload.pull_request;
  let action = payload.action as string;

  let shouldRenderBody = true;
  if (['closed', 'edited'].includes(action)) {
    shouldRenderBody = false;
  }

  const builder = new StringBuilder();

  if (action === 'closed') {
    const pr = (payload as ExtractPayload<'pull_request.closed'>).pull_request;
    if (pr.merged) {
      // If the action is closed and the merged key is true, the pull request was merged.
      action = 'merged';
    }
  }

  // ready_for_review -> ready for review
  action = prettyUnderlineWord(action);

  let shouldRenderMergeInfo = false;
  if (['opened', 'edited', 'merged'].includes(action)) {
    shouldRenderMergeInfo = true;
  }

  const base = (data as PullRequest).base;

  let oldTitle = '';
  let oldRef = '';
  if (action === 'edited') {
    if ((payload as THasChanges).changes) {
      const changes = (payload as THasChanges).changes;
      if (changes?.title) {
        // 说明是标题改变
        // 我们只接收标题带有 WIP 的改变
        if (
          changes.title.from.toLowerCase().includes('wip') &&
          !data.title.toLowerCase().includes('wip')
        ) {
          oldTitle = changes.title.from;
        }
        if (
          !changes.title.from.toLowerCase().includes('wip') &&
          data.title.toLowerCase().includes('wip')
        ) {
          oldTitle = changes.title.from;
        }
      }
      if (changes?.base) {
        oldRef = changes.base.ref.from;
      }
    }

    if (!oldRef && !oldTitle) {
      throw new StopHandleError('ignore pr content change');
    }
  }

  if (oldTitle) {
    builder.add(
      DeletedPrOrIssueTitleLink({
        ...data,
        title: oldTitle,
      }),
    );
  }

  builder.add(IssuesTitleLink(data));

  if (shouldRenderMergeInfo) {
    builder.add(`> ${PullRequestRefInfo(data)}`);
  }

  if (data.requested_reviewers?.length) {
    builder.add(`> ${RequestedReviewersInfo(data.requested_reviewers)}`);
  }

  if (data.assignees?.length) {
    builder.add(`> ${AssigneesInfo(data.assignees)}`);
  }

  if (oldRef) {
    builder.add(
      `> changed the base branch from \`${oldRef}\` to \`${base.ref}\`  `,
    );
  }

  return Template(
    {
      payload,
      event: `${nameBlock}#${data.number}`,
      action,
      title: `{{sender | link:sender}} ${action} [${nameBlock}](${data.html_url})`,
      target: builder.build(),
      body: data.body,
      contentLimit: ctx.setting.contentLimit,
      doNotRenderBody: !shouldRenderBody,
    },
    ctx,
  );
}

export async function handleIssue(
  payload: ExtractPayload<'issues'>,
  ctx: Context,
): Promise<TemplateRenderResult> {
  return render('issues', payload, payload.issue, ctx);
}

export async function handleDiscussion(
  payload: ExtractPayload<'discussion'>,
  ctx: Context,
): Promise<TemplateRenderResult> {
  return render('discussion', payload, payload.discussion, ctx);
}
