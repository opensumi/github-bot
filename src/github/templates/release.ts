import { StringBuilder } from '@/utils';

import { Context , ExtractPayload } from '../types';

import { titleTpl } from './utils';

import { renderUserLink, useRef, textTpl } from '.';

export async function handleRelease(
  payload: ExtractPayload<'release'>,
  ctx: Context,
) {
  const action = payload.action;
  const release = payload.release;
  const repo = payload.repository;
  const title = titleTpl(
    {
      repo,
      event: release.name,
      action,
    },
    ctx,
    false,
  );

  const builder = new StringBuilder();

  let status = '';
  if (release.draft) {
    status = 'Draft';
  }
  if (release.prerelease) {
    status = 'Pre Release';
  }

  if (status) {
    builder.add(`Status: ${status}\n`);
  }

  builder.add(`Tag: ${release.tag_name}\n`);
  builder.add(`>\n${useRef(release.body)}`);

  const text = textTpl(
    {
      repo: payload.repository,
      title: `${renderUserLink(payload.sender)} ${action} [${release.name}](${
        release.html_url
      })`,
      body: builder.build(),
    },
    ctx,
  );

  return { title, text };
}
