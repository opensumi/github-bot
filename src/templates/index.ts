export * from './prOrIssue';
export * from './utils';
