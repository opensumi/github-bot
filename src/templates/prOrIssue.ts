import type { ExtractPayload } from '../template';
import {
  renderRepoLink,
  renderUserLink,
  renderPrOrIssue,
  renderPrOrIssueLink,
} from './utils';
import { Issue, PullRequest } from '@octokit/webhooks-types';

type Name = 'issues' | 'pull_request';
const NameBlock = {
  issues: 'Issue',
  pull_request: 'Pull Request',
} as {
  [key in Name]: string;
};

function render(
  name: Name,
  payload: ExtractPayload<Name>,
  data: PullRequest | Issue,
) {
  const nameBlock = NameBlock[name];
  const action = payload.action.replaceAll('_', '');
  const sender = renderUserLink(payload.sender);
  const originAuthor = renderUserLink(data.user);

  let shouldRenderOriginAuthor = false;
  if (action !== 'opened') {
    shouldRenderOriginAuthor = true;
  }

  const title = `${nameBlock} ${renderPrOrIssueLink(
    data,
  )} ${action} by ${sender}`;

  const text = `${renderRepoLink(
    payload.repository,
  )} ${nameBlock} ${action} by ${renderUserLink(payload.sender)}${
    shouldRenderOriginAuthor ? `, author: ${originAuthor}` : ''
  }

${renderPrOrIssue(data)}`;

  return {
    title,
    text,
  };
}

export function handlePr(payload: ExtractPayload<'pull_request'>) {
  return render('pull_request', payload, payload.pull_request);
}

export function handleIssue(payload: ExtractPayload<'issues'>) {
  return render('issues', payload, payload.issue);
}
