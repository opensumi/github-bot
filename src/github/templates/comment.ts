import {
  limitLine,
  renderPrOrIssueLink,
  renderPrOrIssueText,
  renderRepoLink,
  renderUserLink,
  useRef,
} from '.';
import { ExtractPayload } from '@/github/types';
import { Issue, PullRequest, Discussion, User } from '@octokit/webhooks-types';

type Name = 'issues' | 'pull_request' | 'discussion';
const NameBlock = {
  issues: 'issue',
  pull_request: 'pull request',
  discussion: 'discussion',
} as {
  [key in Name]: string;
};

const formatByUserLogin = {
  'codecov-commenter': (text: string) => {
    return limitLine(text, 3);
  },
} as {
  [key: string]: (text: string) => string;
};

function renderCommentBody(comment: { body: string; user: User }) {
  let text = useRef(comment.body);
  const formatter = formatByUserLogin[comment.user.login];
  if (formatter) {
    text = formatter(text);
  }
  return text;
}

function renderComment(
  name: Name,
  payload: ExtractPayload<'issue_comment' | 'discussion_comment'>,
  data: Issue | PullRequest | Discussion,
) {
  const comment = payload.comment;
  const location = NameBlock[name];
  const action = payload.action;
  let shouldRenderBody = true;
  if (['edited'].includes(action)) {
    shouldRenderBody = false;
  }

  const title = `[${
    payload.repository.name
  }] comment ${action} on ${location} ${renderPrOrIssueText(data)}`;
  const text = `${renderRepoLink(payload.repository)} ${renderUserLink(
    payload.sender,
  )} ${action} [comment](${
    comment.html_url
  }) on ${location} ${renderPrOrIssueLink(data)}${
    shouldRenderBody ? `\n>\n${renderCommentBody(payload.comment)}` : ''
  }
`;
  return {
    title,
    text,
  };
}

export function handleIssueComment(payload: ExtractPayload<'issue_comment'>) {
  const issue = payload.issue;
  const isUnderPullRequest = Boolean(issue.pull_request);
  let name = 'issues' as Name;
  if (isUnderPullRequest) {
    name = 'pull_request';
  }
  return renderComment(name, payload, issue);
}

export function handleDiscussionComment(
  payload: ExtractPayload<'discussion_comment'>,
) {
  return renderComment('discussion', payload, payload.discussion);
}

export function handleCommitComment(payload: ExtractPayload<'commit_comment'>) {
  const comment = payload.comment;
  const commitId = comment.commit_id.slice(0, 6);
  const action = payload.action;

  const title = `[${payload.repository.name}] commit comment ${action} on commit@${commitId}`;
  const text = `${renderRepoLink(payload.repository)} ${renderUserLink(
    payload.sender,
  )} ${action} [commit comment](${comment.html_url}) on [${
    payload.repository.name
  }@${commitId}](${comment.html_url})
>\n
${renderCommentBody(payload.comment)}
`;
  return {
    title,
    text,
  };
}
