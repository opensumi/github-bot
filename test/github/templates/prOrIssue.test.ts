import { handlePr, handleIssue } from '@/github/templates/prOrIssue';

import {
  pull_request_closed,
  pull_request_opened,
  pull_request_13_opened,
  pull_request_edited_wip,
  pull_request_edited_base,
  issue_opened_event,
} from '../../fixtures';
import { ctx } from '../ctx';

describe('github templates pr or issue', () => {
  it('can handle pull_request_closed', async () => {
    const result = await handlePr(pull_request_closed, ctx);
    console.log(`pull_request_closed ~ result`, result);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.title).toBeDefined();
  });
  it('can handle pull_request_opened', async () => {
    const result = await handlePr(pull_request_opened, ctx);
    console.log(`pull_request_opened ~ result`, result.text);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.text).toContain('<-');
    expect(result.title).toBeDefined();
  });
  it('can handle pull_request_edit wip', async () => {
    const result = await handlePr(pull_request_edited_wip, ctx);
    console.log(`pull_request_edited_wip ~ result`, result.text);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.text).toContain('<-');
    expect(result.text).toContain('~~');
    expect(result.title).toBeDefined();
  });
  it('can handle pull_request_edited_base', async () => {
    const result = await handlePr(pull_request_edited_base, ctx);
    console.log(`pull_request_edited_base ~ result`, result.text);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.text).toContain('<-');
    expect(result.text).toContain('> changed the base branch');
    expect(result.title).toBeDefined();
  });
  it('can handle issue', async () => {
    const result = await handleIssue(issue_opened_event, ctx);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.title).toBeDefined();
  });
  it('can handle pull_request_13_opened', async () => {
    const result = await handlePr(pull_request_13_opened, ctx);
    console.log(`pull_request_13_opened ~ result`, result.text);
    expect(result).toMatchSnapshot();
    expect(result.text).toBeDefined();
    expect(result.text).toContain('Requested reviewers');
    expect(result.title).toBeDefined();
  });
});
