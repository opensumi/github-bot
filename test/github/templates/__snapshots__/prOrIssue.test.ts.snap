// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github templates pr or issue can handle issue 1`] = `
{
  "compactText": undefined,
  "text": "#### [bytemain](https://github.com/bytemain) opened [issue](https://github.com/opensumi/core/issues/2960)  
#### [#2960 [BUG]  Problem 面板闪烁](https://github.com/opensumi/core/issues/2960)
> Assignees: [bytemain](https://github.com/bytemain)  

***

打开 settings.json，打开问题面板，操作的时候发现有闪烁现象。

看起来是不同的插件给某个文件做 diagnose 时，传的 uri 没有合并起来？

https://github.com/opensumi/core/assets/1/8eb4bd97-5594-4428-9000-915546e33e63",
  "title": "Issue#2960 opened",
}
`;

exports[`github templates pr or issue can handle pull_request_13_opened 1`] = `
{
  "compactText": undefined,
  "text": "#### [Codertocat](https://github.com/Codertocat) opened [pull request](https://github.com/Codertocat/Hello-World/pull/2)  
#### [#2 Update the README with new information.](https://github.com/Codertocat/Hello-World/pull/2)
> master <- changes  
> Requested reviewers: [octocat](http://github.com/octocat)  
> Assignees: [Codertocat](https://github.com/Codertocat)  ",
  "title": "Pull request#2 opened",
}
`;

exports[`github templates pr or issue can handle pull_request_closed 1`] = `
{
  "compactText": undefined,
  "text": "#### [Codertocat](https://github.com/Codertocat) closed [pull request](https://github.com/Codertocat/Hello-World/pull/2)  
#### [#2 Update the README with new information.](https://github.com/Codertocat/Hello-World/pull/2)
> Requested reviewers: [octocat](http://github.com/octocat)  
> Assignees: [Codertocat](https://github.com/Codertocat)  ",
  "title": "Pull request#2 closed",
}
`;

exports[`github templates pr or issue can handle pull_request_edit wip 1`] = `
{
  "compactText": undefined,
  "text": "#### [bytemain](https://github.com/bytemain) edited [pull request](https://github.com/opensumi/core/pull/1235)  
#### [~~#1235 ci: add codeql-analysis.yml~~](https://github.com/opensumi/core/pull/1235)
#### [#1235 wip: ci: add codeql-analysis.yml](https://github.com/opensumi/core/pull/1235)
> main <- bytemain-patch-2  ",
  "title": "Pull request#1235 edited",
}
`;

exports[`github templates pr or issue can handle pull_request_edited_base 1`] = `
{
  "compactText": undefined,
  "text": "#### [bytemain](https://github.com/bytemain) edited [pull request](https://github.com/opensumi/core/pull/1235)  
#### [#1235 wip: ci: add codeql-analysis.yml](https://github.com/opensumi/core/pull/1235)
> alipay/v2.18 <- bytemain-patch-2  
> changed the base branch from \`main\` to \`alipay/v2.18\`  ",
  "title": "Pull request#1235 edited",
}
`;

exports[`github templates pr or issue can handle pull_request_opened 1`] = `
{
  "compactText": undefined,
  "text": "#### [Codertocat](https://github.com/Codertocat) opened [pull request](https://github.com/Codertocat/Hello-World/pull/2)  
#### [#2 Update the README with new information.](https://github.com/Codertocat/Hello-World/pull/2)
> master <- changes  

***

This is a pretty simple change that we need to pull into master.",
  "title": "Pull request#2 opened",
}
`;
